lightning/src/onion_message/messenger.rs

```diff
@@ -406,51 +485,18 @@ where
 	>(
 		&self, recipient: PublicKey, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
 	) -> Result<Vec<BlindedPath>, ()> {
-		// Limit the number of blinded paths that are computed.
-		const MAX_PATHS: usize = 3;
-
-		// Ensure peers have at least three channels so that it is more difficult to infer the
-		// recipient's node_id.
-		const MIN_PEER_CHANNELS: usize = 3;
-
-		let network_graph = self.network_graph.deref().read_only();
-		let is_recipient_announced =
-			network_graph.nodes().contains_key(&NodeId::from_pubkey(&recipient));
-
-		let mut peer_info = peers.iter()
-			// Limit to peers with announced channels
-			.filter_map(|pubkey|
-				network_graph
-					.node(&NodeId::from_pubkey(pubkey))
-					.filter(|info| info.channels.len() >= MIN_PEER_CHANNELS)
-					.map(|info| (*pubkey, info.is_tor_only(), info.channels.len()))
-			)
-			// Exclude Tor-only nodes when the recipient is announced.
-			.filter(|(_, is_tor_only, _)| !(*is_tor_only && is_recipient_announced))
-			.collect::<Vec<_>>();
-
-		// Prefer using non-Tor nodes with the most channels as the introduction node.
-		peer_info.sort_unstable_by(|(_, a_tor_only, a_channels), (_, b_tor_only, b_channels)| {
-			a_tor_only.cmp(b_tor_only).then(a_channels.cmp(b_channels).reverse())
-		});
-
-		let paths = peer_info.into_iter()
-			.map(|(pubkey, _, _)| vec![pubkey, recipient])
-			.map(|node_pks| BlindedPath::new_for_message(&node_pks, &*self.entropy_source, secp_ctx))
-			.take(MAX_PATHS)
-			.collect::<Result<Vec<_>, _>>();
+		let peers = peers
+			.into_iter()
+			.map(|node_id| ForwardNode { node_id, short_channel_id: None });
+		self.create_blinded_paths_from_iter(recipient, peers, secp_ctx, false)
+	}
 
-		match paths {
-			Ok(paths) if !paths.is_empty() => Ok(paths),
-			_ => {
-				if is_recipient_announced {
-					BlindedPath::one_hop_for_message(recipient, &*self.entropy_source, secp_ctx)
-						.map(|path| vec![path])
-				} else {
-					Err(())
-				}
-			},
-		}
+	fn create_compact_blinded_paths<
```

Comment:
We can't really have the implementation pick without parameterizing it with a `ShortChannelIdLookUp`, which we removed in the #3011 (see https://github.com/lightningdevkit/rust-lightning/pull/3011#discussion_r1595871667). Also, once we have three-hop blinded paths, the implementation *will* need to make the decision for the additional hop using a `NetworkGraph`, as it does for the introduction node. Do you prefer re-introducing `ShortChannelIdLookUp` so that it is all in one place?

Reply:
> We could go the other way, right? Always create a non-compact path (just do whatever the hops passed say to do) and then have the caller compact the path if they want. That doesn't allow us to communicate to the `MessageRouter` that we want a fewer-hop-count path, though.

For the introduction node, the caller can't compact if they don't have a `NetworkGraph` to use, which `ChannelManager` doesn't directly (only via `DefaultRouter`). For intermediate nodes, the caller can't compact after the path is already built since the scid is in the encrypted data. And for three-hop paths, the same `NetworkGraph` issues arises.

> 
> I guess maybe the variation between `create_blinded_paths` taking `PublicKey`s and `create_compact_blinded_paths` taking `ForwardNode`s is maybe enough, it just seems weird that the scids in `ForwardNode`s are `Option`al in that API.

Yeah, I think that may be the most straight-forward option. As for the `Option`al scid in `ForwardNode`, it makes degrading to a partial compact path (i.e., only introduction node compact) simple. See https://github.com/lightningdevkit/rust-lightning/pull/3011#discussion_r1612286686.